File: data_structure/ruby/priority-queue/linked-list/pqueue.rb
# PQueue using Linked Lists

class Node
  attr_accessor :data, :priority, :next

  def initialize(data, priority)
    @data = data
    @priority = priority
    @next = nil
  end
end

class PQueue
  def initialize
    @root = nil
  end

  def push(data, priority)
    new_node = Node.new(data, priority)

    if @root == nil
      @root = new_node
    elsif @root.priority < priority
      new_node.next = @root
      @root = new_node
    else
      # traverse the list and find the proper position
      curr = @root
      while curr.next != nil and curr.next.priority < priority
        curr = curr.next
      end

      new_node.next = curr.next
      curr.next = new_node
    end
  end

  # removes the highest priority element
  def pop
    return if @root == nil
    data = @root.data
    @root = @root.next

    puts "pop #{data}"
    data
  end

  # get highest priority element
  def peek
    @root.data
  end

  def is_empty?
    return @root==nil
  end

  def print_list
    return if @root == nil
    curr = @root
    while curr.next != nil
      print "(#{curr.data}, #{curr.priority}) -> "
      curr = curr.next
    end
    puts "(#{curr.data}, #{curr.priority})"
  end
end

pq = PQueue.new()
0.upto(5) do |x|
  pq.push((97+x).chr, 97+x)
  pq.print_list
  puts
end

pq.pop
pq.print_list
pq.pop
pq.print_list
pq.pop
pq.print_list
pq.pop
pq.print_list
pq.pop
pq.print_list
pq.pop
pq.print_list
pq.pop
pq.print_list

File: bin/old_ds/cpp/priority-queue/linkedlist/pqueue.c
#include "pqueue.h"

/* constructor */
Node* NewNode(const Element data, const int priority)
{
	Node *work = malloc(sizeof(*work));
	work->data = data;
	work->priority = priority;

	return work;
}

PQueue* NewPriorityQueue(void)
{
	PQueue *pq = malloc(sizeof(*pq));
	pq->head = NewNode(DATA, PRIORITY);
	pq->tail = NewNode(DATA, PRIORITY);
	pq->head->next = pq->tail;
	pq->tail->prev = pq->head;
	pq->size = 0;

	return pq;
}	

/* destructor */
void DestroyList(PQueue *pq)
{
	Node *head = pq->head->next;
	Node *end = pq->tail;

	while(head != end)
	{
		Node *temp = head->next;
		free(head);
		head = temp;
	}

	free(pq->head);
	free(end);
	free(pq);
}

/* list modifier */
void Push(PQueue *pq, const Element data, const int priority)
{
	Node *curr = pq->head->next;
	Node *end = pq->tail;

	for(; curr != end; curr = curr->next)
	{
		// highest priority comes first
		if(curr->priority < priority)
		{
			break;
		}
	}

	Node *work = NewNode(data, priority);
	ConnectNode(work, curr);
	pq->size++;
}

void Pop(PQueue *pq)
{
	if(pq->size > 0)
	{
		printf("deleting the last node..\n");
		Delete(pq, pq->tail->prev);
	}
}

void DeleteNthNode(PQueue *pq, int n)
{
	if(n <= 0 || n > pq->size)
	{
		fprintf(stderr, "out of bound...\n");
		return;
	}

	printf("deleting node #%d..\n", n);
	Node *work = pq->head->next;

	for(int i=0; i<n-1; ++i)
	{
		work = work->next;
	}

	Delete(pq, work);
}

/* list access */
void Print(const PQueue *pq)
{
	Node *beg = pq->head->next;
	Node *end = pq->tail;

	while(beg->next != end)
	{
		if(beg == end)
		{
			break;
		}

		printf("(%d, %d) -> ", beg->data, beg->priority);
		beg = beg->next;
	}
	printf("(%d, %d)\n", beg->data, beg->priority);
}

bool IsEmpty(const PQueue *pq)
{
	return pq->size == 0;
}

int GetSize(const PQueue *pq)
{
	return pq->size;
}

Element GetHighestPriority(const PQueue *pq)
{
	if(pq->size <= 0)
	{
		fprintf(stderr, "list is empty...\n");
		return -1;
	}

	return pq->head->next->data;
}

Element GetSmallestPriority(const PQueue *pq)
{
	if(pq->size <= 0)
	{
		fprintf(stderr, "list is empty...\n");
		return -1;
	}

	return pq->tail->prev->data;
}

/* helper functions */
static void ConnectNode(Node *work, Node *curr)
{
	curr->prev->next = work;
	work->prev = curr->prev;
	curr->prev = work;
	work->next = curr;
}

static void Delete(PQueue *pq, Node *remove)
{
	DisconnectNode(remove);
	free(remove);
	remove = NULL;
	pq->size--;
}

static void DisconnectNode(Node *curr)
{
	curr->prev->next = curr->next;
	curr->next->prev = curr->prev;
}
